pkg/ottl/ottlfuncs/func_parse_key_value.go

```diff
@@ -0,0 +1,81 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package ottlfuncs
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/coreinternal/parseutils"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+)
+
+type ParseKeyValueArguments[K any] struct {
+	Target        ottl.StringGetter[K]
+	Delimiter     ottl.Optional[string]
+	PairDelimiter ottl.Optional[string]
+}
+
+func NewParseKeyValueFactory[K any]() ottl.Factory[K] {
+	return ottl.NewFactory("ParseKeyValue", &ParseKeyValueArguments[K]{}, createParseKeyValueFunction[K])
+}
+
+func createParseKeyValueFunction[K any](_ ottl.FunctionContext, oArgs ottl.Arguments) (ottl.ExprFunc[K], error) {
+	args, ok := oArgs.(*ParseKeyValueArguments[K])
+
+	if !ok {
+		return nil, fmt.Errorf("ParseKeyValueFactory args must be of type *ParseKeyValueArguments[K]")
+	}
+
+	return parseKeyValue[K](args.Target, args.Delimiter, args.PairDelimiter)
+}
+
+func parseKeyValue[K any](target ottl.StringGetter[K], d ottl.Optional[string], p ottl.Optional[string]) (ottl.ExprFunc[K], error) {
+	delimiter := "="
+	if !d.IsEmpty() && d.Get() != "" {
+		delimiter = d.Get()
+	}
+
+	pairDelimiter := " "
+	if !p.IsEmpty() && p.Get() != "" {
+		pairDelimiter = p.Get()
+	}
+
+	if pairDelimiter == delimiter {
+		return nil, fmt.Errorf("pair delimiter %q cannot be equal to delimiter %q", pairDelimiter, delimiter)
+	}
+
+	return func(ctx context.Context, tCtx K) (any, error) {
+		source, err := target.Get(ctx, tCtx)
+		if err != nil {
+			return nil, err
+		}
+
+		if source == "" {
+			return nil, fmt.Errorf("cannot parse from empty target")
+		}
+
+		pairs, err := parseutils.SplitString(source, pairDelimiter)
```

Comment:
Can we emit a debug/warning log if the pairDelimiter isn't found? I think this is something we would want to alert the operator to while still trying to emit what data we can. You can get a logger from `ottl.FunctionContext`.

@TylerHelmuth What are your thoughts on this? I think it would make sense to emit what data we can, but maybe we should just return an error here. I don't think there's much precedence for this in OTTL, usually we just return an error.

Reply:
A debug may be best because in the case of malformed input the results of further parsing will be totally broken, and it may not be totally obvious why. It could be as simple as, "no pair delimiter detected, possible malformed input" or something of the like.

Reply:
Special cases aside, I think it's very common that users will just want the default values for both delimiter and pair delimiter. I wouldn't emit warnings if using the defaults.

Reply:
@evan-bradley I'm not really clear on the concern.  I don't see stanza handling that situation as a special case so I don't think we need to either.